### PR TITLE
Scrapper, Egor Dzyumenko - 21FPL1

### PR DESCRIPTION
**Pull request by EgDzum**

### CODE REVIEW

File: lab_5_scrapper/scrapper.py [hunk 1]
```diff
@@ -13,70 +46,108 @@ def __init__(self, path_to_config: Path) -> None:
...
         Initializes an instance of the Config class
         """
-        pass
+        self.seed_urls = path_to_config.seed_urls
+        self.total_articles_to_find_and_parse = path_to_config.total_articles_to_find_and_parse
+        self.headers = path_to_config.headers
+        self.encoding = path_to_config.encoding
+        self.timeout = path_to_config.timeout
+        self.verify_certificate = path_to_config.verify_certificate
+        self.headless_mode = path_to_config.headless_mode
```
Comment (dmitry-uraev) [inline lab_5_scrapper/scrapper.py]: You define getting methods for all these class fields, but why should I use these getting methods? I can simply call smth like `class_instance.headers` and get headers string and there is no need in `class_instance.get_headers()` method. 

Think about OOP principles, e.g. encapsulation, and fix you realisation accordingly, so I could use only getting methods.

**Review by dmitry-uraev — CHANGES_REQUESTED** [review]
Good start! Please, continue implementing lab 5.

File: lab_5_scrapper/scrapper.py [hunk 2]
```diff
@@ -90,25 +196,44 @@ def __init__(self, config: Config) -> None:
...
+                self.urls.append(link)
+                if len(self.urls) == self.config.get_num_articles():
+                    break
 
     def get_search_urls(self) -> list:
         """
         Returns seed_urls param
         """
-        pass
+        return self.config.get_seed_urls()
```
Comment (dmitry-uraev) [inline lab_5_scrapper/scrapper.py]: why cant i just write 

``crawler.config.get_seed_urls()``
    
   Why should i use ``crawler.get_search_urls()`` method instead?

Comment (dmitry-uraev) [reply lab_5_scrapper/scrapper.py]: And you also return seed_urls, but method returns search_urls - different variables

File: lab_5_scrapper/scrapper.py [hunk 3]
```diff
@@ -120,21 +245,31 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
+        self.article.text = text
 
     def _fill_article_with_meta_information(self, article_soup: BeautifulSoup) -> None:
         """
         Finds meta information of article
         """
         pass
 
-    def unify_date_format(self, date_str: str) -> datetime.datetime:
+    def unify_date_format(self, date_str: str):
```
Comment (dmitry-uraev) [inline lab_5_scrapper/scrapper.py]: Do not modify type definitions

**Review by dmitry-uraev — CHANGES_REQUESTED** [review]
There are some comments to fix

File: lab_5_scrapper/scrapper.py [hunk 4]
```diff
@@ -90,25 +198,51 @@ def __init__(self, config: Config) -> None:
         """
         Initializes an instance of the Crawler class
         """
-        pass
+        self._config = config
+        self._seed_urls = self._config.get_seed_urls()
+        self.urls = []
 
-    def _extract_url(self, article_bs: BeautifulSoup) -> str:
+    @staticmethod
```
Comment (dmitry-uraev) [inline lab_5_scrapper/scrapper.py]: why do you need this?

File: lab_5_scrapper/scrapper.py [hunk 5]
```diff
@@ -90,25 +198,51 @@ def __init__(self, config: Config) -> None:
...
+    def _extract_url(article_bs: BeautifulSoup) -> str:
         """
         Finds and retrieves URL from HTML
         """
-        pass
+        href = article_bs.get("href")
+
+        if href and href.startswith("https://irkutskmedia.ru/news/") and 'hashtag' not in href:
```
Comment (dmitry-uraev) [inline lab_5_scrapper/scrapper.py:211]: can you simplify this condition check?

File: lab_5_scrapper/scrapper.py [hunk 6]
```diff
@@ -120,45 +254,129 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
+        date = article_soup.find('div', class_='fn-rubric-link')
+
+        if date:
+            self.article.date = self.unify_date_format(date.text.strip())
+
+        # the name of authors is not defined on the pages
+        self.article.author = ["NOT FOUND"]
 
-    def unify_date_format(self, date_str: str) -> datetime.datetime:
+    @staticmethod
```
Comment (dmitry-uraev) [inline lab_5_scrapper/scrapper.py]: again staticmethod, why do you use that?

File: lab_5_scrapper/scrapper.py [hunk 7]
```diff
@@ -120,45 +254,129 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
+
+        # 25 декабря, 11:30
+        # the pattern is aimed to find date information like in the example above(without year)
+        if re.match(r'\d{2} (?i)[a-z]+, \d{2}:\d{2}', eng_date):
+            date_d = datetime.datetime.strptime(eng_date, '%d %B, %H:%M')
+            return date_d.replace(year=this_year)
+
+        # 25 декабря 2023, 11:30
+        # matches when a parsed value has date, month, year and time
+        return datetime.datetime.strptime(eng_date, '%d %B %Y, %H:%M')
```
Comment (dmitry-uraev) [inline lab_5_scrapper/scrapper.py]: a lot of code and logic - can you simplify date extraction?

**Review by dmitry-uraev — CHANGES_REQUESTED** [review]
Good! Do not forget to respond to my comments

**Review by demid5111 — CHANGES_REQUESTED** [review]
https://irkutskmedia.ru/news/1487750/?from=50 - too many text (keep only article's text)

File: lab_5_scrapper/scrapper.py (hunk 4 above)
Comment (EgDzum) [reply lab_5_scrapper/scrapper.py]: removed

File: lab_5_scrapper/scrapper.py (hunk 6 above)
Comment (EgDzum) [reply lab_5_scrapper/scrapper.py]: removed

File: lab_5_scrapper/scrapper.py (hunk 3 above)
Comment (EgDzum) [reply lab_5_scrapper/scrapper.py]: type returned

File: lab_5_scrapper/scrapper.py [hunk 8]
```diff
@@ -120,45 +253,151 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
+
+        # 25 декабря, 11:30
+        # the pattern is aimed to find date information like in the example above(without year)
+        if re.match(r'\d{2} \w+, \d{2}:\d{2}', date_str):
+            date_d = datetime.datetime.strptime(eng_date, '%d %B, %H:%M')
+            return date_d.replace(year=this_year)
+
+        # 25 декабря 2023, 11:30
+        # matches when a parsed value has date, month, year and time
+        return datetime.datetime.strptime(eng_date, '%d %B %Y, %H:%M')
```
Comment (dmitry-uraev) [inline lab_5_scrapper/scrapper.py:359]: is there a way to simplify logic in this really large method?

**Review by dmitry-uraev — CHANGES_REQUESTED** [review]
one more comment to respond to and we are done

File: lab_5_scrapper/scrapper.py (hunk 8 above)
Comment (EgDzum) [reply lab_5_scrapper/scrapper.py:359]: Basically, there are 4 types of time representation that we can come across on the site, for example:
1. 11:30
2. 21.04.2023
3. 25 декабря, 11:30
4. 25 декабря 2023, 11:30

Considering different regex patterns as well as patterns from datetime library for each case, we need to make up 4 different conditions to be able to parse the date from any article.

**Review by dmitry-uraev — APPROVED** [review]

**Pull request closed without merging**